### PR TITLE
DAPI-347: Initial commit notification sample app

### PR DESCRIPTION
**Pull request by mitel-prashant**
Resolves [DAPI-347](https://mitel.atlassian.net/browse/DAPI-347)

This PR fix DAPI-347 - This sample code makes use of the Create Subscription endpoint to create a subscription for ringing events on a range of endpoints defined by the user. This sample code is expected to output any received webhooks to the console.

### CODE REVIEW

File: sample-app/notification-sample-code/nodejs/index.js [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+    try {
+        // Add client_id in userData to send in get code api request body
+        userData.client_id = inputData.appData.client_id;
+        // Encode form data in URL encoded format and pass it request body of token api
+        let formData = Object.keys(userData).map((key) => {
+            return encodeURIComponent(key) + '=' + encodeURIComponent(userData[key]);
+        }).join('&');
+
+        // Call the code api to obtain the valid code
+        const response = await fetch('https://authentication.dev.api.mitel.io/2017-09-01/authorize', {
```
Comment (FerrissK) [inline sample-app/notification-sample-code/nodejs/index.js]: This auth call should point to the public cloud

File: sample-app/notification-sample-code/nodejs/index.js [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+ async function getCode(userData) {
+    try {
+        // Add client_id in userData to send in get code api request body
+        userData.client_id = inputData.appData.client_id;
+        // Encode form data in URL encoded format and pass it request body of token api
+        let formData = Object.keys(userData).map((key) => {
+            return encodeURIComponent(key) + '=' + encodeURIComponent(userData[key]);
+        }).join('&');
+
+        // Call the code api to obtain the valid code
```
Comment (FerrissK) [inline sample-app/notification-sample-code/nodejs/index.js]: This should read: // Call the auth api to obtain a valid code

File: sample-app/notification-sample-code/nodejs/index.js [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+/**
+ * Main function is entry point for application
+ * @returns Valid caller id
+ */
+async function main() {
+    // Obtain token for User 
+    // Call the getCode and getToken function to obtain the valid token
+    const userCode = await getCode(inputData.userdata);
+    const token = await getToken(userCode.code, inputData.appData.client_id);
```
Comment (FerrissK) [inline sample-app/notification-sample-code/nodejs/index.js:164]: Should validate userCode before using in the getToken

File: sample-app/notification-sample-code/nodejs/index.js [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    try {
+        // Create form data for authorize api to fetch bearer token
+        let formData = {
+            grant_type: "authorization_code",
+            code: code,
+            client_id: clientId
+        }
+
+        // Call the token api to obtain the valid token
+        const response = await fetch('https://authentication.dev.api.mitel.io/2017-09-01/token', {
```
Comment (FerrissK) [inline sample-app/notification-sample-code/nodejs/index.js]: Should call public cloud

File: sample-app/notification-sample-code/nodejs/index.js [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+/**
+ * Function to create subscription on ringing event
+ * @param {access token} token 
+ * @returns 
+ */
+async function createSubscription(token) {
+    try {
+        // Call create conversation API to start the conversation
+        const response = await fetch('https://notifications.dev.api.mitel.io/2017-09-01/subscriptions', {
```
Comment (FerrissK) [inline sample-app/notification-sample-code/nodejs/index.js]: Should call public cloud

File: sample-app/notification-sample-code/nodejs/index.js [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+ */
+async function makeCall(token) {
+    try {
+        // Get the generic messga body from input data
+        let callBody = {
+            to: inputData.makeCallData.to
+        }
+
+        // Call the message API to send message
+        const response = await fetch(`https://media.dev.api.mitel.io/2017-09-01/endpoints/${inputData.makeCallData.from}/calls`, {
```
Comment (FerrissK) [inline sample-app/notification-sample-code/nodejs/index.js]: Should call public cloud

**Review by FerrissK — CHANGES_REQUESTED** [review]

File: sample-app/notification-sample-code/nodejs/README.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
+# Notification Sample Code
+
+This sample code makes use of the Create Subscription endpoint to create a subscription for ringing events on a range of endpoints defined by the user. This sample code is expected to output any received webhooks to the console
+
+## Pre-requisites
+
+[Node.js](https://nodejs.org/en/) - Install node.js
```
Comment (FerrissK) [inline sample-app/notification-sample-code/nodejs/README.md]: We should probably add having a cloudlink account and obtaining a cliend Id as pre- requisites.

@buddrenaud, maybe we can link or at least tell where they can read your get starting doc?

File: sample-app/notification-sample-code/nodejs/README.md [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+# Provide valid input data
+
+Open input inputData.js file and modify the below details
+
+```json
+{
+    userdata: {
+        response_type: "code",       // no change needed
+        username: "USER_NAME",       // provide your username for user 1
```
Comment (FerrissK) [inline sample-app/notification-sample-code/nodejs/README.md]: We should specify that this is a Cloudlink username/ password

File: sample-app/notification-sample-code/nodejs/README.md [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+        password: "PASSWORD",        // provide password for user
+        account_id: "ACCOUNT_ID",    // provide account_id for user
+    },
+    appData: {
+        client_id: "CLIENT_ID"       // provide valid unique clientId issued by Mitel
+    },
+    // Data for subscription creation
+    subscriptionData: {
+        applicationId: "APPLICATION_ID", // 128-bit GUID that identifies the application for trust/signing purposes
```
Comment (FerrissK) [inline sample-app/notification-sample-code/nodejs/README.md]: I got confused over what an APPLICATION_ID was, we should be more specific that this is an ID that you specify.

File: sample-app/notification-sample-code/nodejs/README.md [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+        deviceId: "DEVICE_ID", //URL for webhooks to be delivered to
```
Comment (FerrissK) [inline sample-app/notification-sample-code/nodejs/README.md]: Add https://webhook.site/ as an example?

File: sample-app/notification-sample-code/nodejs/README.md [hunk 11]
```diff
@@ -23,17 +25,17 @@ Open input inputData.js file and modify the below details
...
         account_id: "ACCOUNT_ID",    // provide account_id for user
     },
     appData: {
         client_id: "CLIENT_ID"       // provide valid unique clientId issued by Mitel
     },
     // Data for subscription creation
     subscriptionData: {
-        applicationId: "APPLICATION_ID", // 128-bit GUID that identifies the application for trust/signing purposes
-        deviceId: "DEVICE_ID", //URL for webhooks to be delivered to
+        applicationId: "APPLICATION_ID", // This is a unique ID specific to your app that’s required if you’re using a webhooks. You can generate this yourself, just make sure it’s unique, so something like a 128-bit GUID is a good choice.
```
Comment (FerrissK) [inline sample-app/notification-sample-code/nodejs/README.md:37]: Change 'using a webhooks' -> 'using a webhook'

File: sample-app/notification-sample-code/nodejs/README.md [hunk 12]
```diff
@@ -4,7 +4,9 @@ This sample code makes use of the Create Subscription endpoint to create a subsc
 
 ## Pre-requisites
 
-[Node.js](https://nodejs.org/en/) - Install node.js
+* Have an cloudlink account and obtain a client id
```
Comment (FerrissK) [inline sample-app/notification-sample-code/nodejs/README.md]: Should we add that you can obtain more information on this through the [Developer Portal](https://developer.mitel.io/)?

File: sample-app/notification-sample-code/nodejs/README.md [hunk 13]
```diff
@@ -23,17 +25,17 @@ Open input inputData.js file and modify the below details
 {
     userdata: {
         response_type: "code",       // no change needed
-        username: "USER_NAME",       // provide your username for user 1
-        password: "PASSWORD",        // provide password for user
+        username: "USER_NAME",       // this is your cloudlink user
+        password: "PASSWORD",        // provide password for your cloudlink user
         account_id: "ACCOUNT_ID",    // provide account_id for user
```
Comment (FerrissK) [inline sample-app/notification-sample-code/nodejs/README.md]: Again, I would add Cloudlink account ID.

File: sample-app/notification-sample-code/nodejs/README.md [hunk 14]
```diff
@@ -23,17 +25,17 @@ Open input inputData.js file and modify the below details
 {
     userdata: {
         response_type: "code",       // no change needed
-        username: "USER_NAME",       // provide your username for user 1
-        password: "PASSWORD",        // provide password for user
+        username: "USER_NAME",       // this is your cloudlink user
```
Comment (FerrissK) [inline sample-app/notification-sample-code/nodejs/README.md]: I would capitalize Cloudlink

File: sample-app/notification-sample-code/nodejs/README.md [hunk 15]
```diff
@@ -23,17 +25,17 @@ Open input inputData.js file and modify the below details
...
+        password: "PASSWORD",        // provide password for your cloudlink user
```
Comment (FerrissK) [inline sample-app/notification-sample-code/nodejs/README.md]: I would capitalize Cloudlink

File: sample-app/notification-sample-code/nodejs/inputData.js [hunk 16]
```diff
@@ -1,17 +1,17 @@
 module.exports = {
     userdata: {
         response_type: "code",       // no change needed
-        username: "USER_NAME",       // provide your username for user 1
-        password: "PASSWORD",        // provide password for user
+        username: "USER_NAME",       // this is your cloudlink user
```
Comment (FerrissK) [inline sample-app/notification-sample-code/nodejs/inputData.js]: Update this file with the comments from the Readme file.

**Review by FerrissK — CHANGES_REQUESTED** [review]

File: sample-app/notification-sample-code/nodejs/index.js (hunk 3 above)
Comment (FerrissK) [reply sample-app/notification-sample-code/nodejs/index.js:164]: Should check userCode for null/ undefined and log out message.

Comment (mitel-prashant) [reply sample-app/notification-sample-code/nodejs/index.js:164]: Below code added in every function will log the exact error coming from API and exit the flow

// If response code is greater than 201, stop the execution and terminate the program
if (response.status > 201) {
	console.log(await response.json());
	process.exit();
}

Comment (FerrissK) [reply sample-app/notification-sample-code/nodejs/index.js:164]: Pefect, I will resolve this conversation.

**Review by FerrissK — APPROVED** [review]

**Pull request merged**